File: invenio_requests/customizations/default/actions.py
# -*- coding: utf-8 -*-
#
# Copyright (C) 2021 TU Wien.
# Copyright (C) 2021 Northwestern University.
#
# Invenio-Requests is free software; you can redistribute it and/or modify
# it under the terms of the MIT License; see LICENSE file for more details.

"""RequestActions define code to be executed when performing actions on requests."""


from invenio_access.permissions import system_process

from ...records.api import RequestEventType
from ..base import RequestAction


class SubmitAction(RequestAction):
    """Submit a request."""

    status_from = ['draft']
    status_to = 'open'


class AcceptAction(RequestAction):
    """Decline a request."""

    status_from = ['open']
    status_to = 'accepted'
    event_type = RequestEventType.ACCEPTED.value


class DeclineAction(RequestAction):
    """Decline a request."""

    status_from = ['open']
    status_to = 'declined'
    event_type = RequestEventType.DECLINED.value


class CancelAction(RequestAction):
    """Cancel a request."""

    status_from = ['open']
    status_to = 'cancelled'
    event_type = RequestEventType.CANCELLED.value


class ExpireAction(RequestAction):
    """Expire a request."""

    status_from = ['open']
    status_to = 'expired'
    event_type = RequestEventType.EXPIRED.value

    def can_execute(self, identity):
        """Check whether the action can be executed."""
        is_system_process = system_process in identity.provides
        return super().can_execute(identity) and is_system_process

File: invenio_requests/assets/semantic-ui/js/invenio_requests/confirmationModal/state/actions.js
// This file is part of InvenioRequests
// Copyright (C) 2022 CERN.
//
// Invenio RDM Records is free software; you can redistribute it and/or modify it
// under the terms of the MIT License; see LICENSE file for more details.
import { errorSerializer } from "../../api/serializers";

export const OPEN_MODAL = "confirmation_modal/OPEN_MODAL";
export const CLOSE_MODAL = "confirmation_modal/CLOSE_MODAL";
export const START_ACTION = "confirmation_modal/START_ACTION";
export const ACTION_SUCCESS = "confirmation_modal/ACTION_SUCCESS";
export const HAS_ERROR = "confirmation_modal/HAS_ERROR";

export const openModal = ({ text, action }) => {
  return async (dispatch) => {
    dispatch({ type: OPEN_MODAL, payload: { text, action } });
  };
};

export const startAction = () => {
  return async (dispatch, getState) => {
    const { modalAction } = getState().confirmationModal;

    try {
      dispatch({ type: START_ACTION });

      await modalAction();

      dispatch({ type: CLOSE_MODAL });
    } catch (error) {
      dispatch({ type: HAS_ERROR, payload: errorSerializer(error) });
    }
  };
};

export const closeModal = () => {
  return async (dispatch) => {
    dispatch({ type: CLOSE_MODAL });
  };
};
